### PR TITLE
[Modules] New module DNS Private Resolvers with new dependency approach

### PR DESCRIPTION
**Pull request by ChrisSidebotham**
# Description

Added new modules for Microsoft.Network/DNSResolvers, added with new dependency approach for Virtual Network/Subnet Requirements.

## Pipeline references

| Pipeline |
| - |
| [Build Status](https://dev.azure.com/carml/CARML/_build/latest?definitionId=116&branchName=users%2Fcsidebotham%2F2199-feature-request-create-new-module-for-private-dns-resolver-service)|

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/workflows/ms.network.dnsresolvers.yml [hunk 1]
```diff
@@ -0,0 +1,145 @@
+name: 'Network: dnsResolvers'
```
Comment (MrMCake) [inline .github/workflows/ms.network.dnsresolvers.yml:1]:
```suggestion
name: 'Network: DNS Resolvers'
```

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+// ========== //
+@description('Optional. The name of the resource group to deploy for testing purposes.')
+@maxLength(90)
+param resourceGroupName string = 'ms.network.dnsResolvers-${serviceShort}-rg'
+
+@description('Optional. The location to deploy resources to.')
+param location string = deployment().location
+
+@description('Optional. A short identifier for the kind of deployment. Should be kept short to not run into resource-name length-constraints.')
+param serviceShort string = 'dnsrcom'
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep:14]:
```suggestion
param serviceShort string = 'ndrcom'
```

Comment (MrMCake) [reply modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep:14]: NOTE: This required you to regenerate the ReadME

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// Test Execution //
+// ============== //
+
+module testDeployment '../../deploy.bicep' = {
+  scope: resourceGroup
+  name: '${uniqueString(deployment().name)}-test-${serviceShort}'
+  params: {
+    name: '<<namePrefix>>${serviceShort}001'
+    virtualNetworkId: resourceGroupResources.outputs.virtualNetworkId
+    // Non-required parameters
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep:46]:
```suggestion
```

File: modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        }
+      }
+    ]
+  }
+}
+
+@description('The resource ID of the created Virtual Network')
+output virtualNetworkId string = virtualNetwork.id
+
+@description('The resource ID of the created Virtual Network Subnet.')
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep:52]:
```suggestion
@description('The resource ID of the created DNS-In Virtual Network Subnet.')
```

Comment (MrMCake) [reply modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep:52]: Not sure if 'incorming' is the correct erm though

File: modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+output subnetResourceId_dnsIn string = virtualNetwork.properties.subnets[0].id
+
+@description('The resource ID of the created Virtual Network Subnet.')
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep:55]:
```suggestion
@description('The resource ID of the created outgoing DNS Virtual Network Subnet.')
```

Comment (MrMCake) [reply modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep:55]: Not sure if 'outgoing' is the correct term though

File: modules/Microsoft.Network/dnsResolvers/.bicep/nested_roleAssignments.bicep [hunk 6]
```diff
@@ -0,0 +1,55 @@
+@sys.description('Required. The IDs of the principals to assign the role to.')
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/.bicep/nested_roleAssignments.bicep:1]: There are a couple parameters missing. Please check e.g the [Logic App module](https://github.com/Azure/ResourceModules/blob/main/modules/Microsoft.Logic/workflows/.bicep/nested_roleAssignments.bicep) for reference.

Also requires an update of the invoking parent bicep.

File: modules/Microsoft.Network/dnsResolvers/deploy.bicep [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+resource dnsResolver_lock 'Microsoft.Authorization/locks@2017-04-01' = if (!empty(lock)) {
+  name: '${dnsResolver.name}-${lock}-lock'
+  properties: {
+    level: any(lock)
+    notes: lock == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: dnsResolver
+}
+
+module dnsResolver_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
```
Comment (MrMCake) [inline modules/Microsoft.Network/dnsResolvers/deploy.bicep:94]: Must be updated as per the suggested additions to the `nested_roleAssignments.bicep` file

File: .github/workflows/ms.network.dnsresolvers.yml (hunk 1 above)
Comment (ChrisSidebotham) [reply .github/workflows/ms.network.dnsresolvers.yml]: Resolved in latest commit

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep (hunk 2 above)
Comment (ChrisSidebotham) [reply modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep]: Resolved in latest commit

File: modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep (hunk 3 above)
Comment (ChrisSidebotham) [reply modules/Microsoft.Network/dnsResolvers/.test/common/deploy.test.bicep]: Resolved in latest commit

File: modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep (hunk 4 above)
Comment (ChrisSidebotham) [reply modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep]: Resolved in latest commit

File: modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep (hunk 5 above)
Comment (ChrisSidebotham) [reply modules/Microsoft.Network/dnsResolvers/.test/common/dependencies.bicep]: Resolved in latest commit

File: modules/Microsoft.Network/dnsResolvers/.bicep/nested_roleAssignments.bicep (hunk 6 above)
Comment (ChrisSidebotham) [reply modules/Microsoft.Network/dnsResolvers/.bicep/nested_roleAssignments.bicep:1]: Resolved in latest commit

**Review by MrMCake — APPROVED** [review]

**Pull request merged**